### PR TITLE
renesas-ra: Add the UART methods uart.txdone() and uart.flush().

### PR DESCRIPTION
**Pull request by robert-hh**
The API is consistent to the other ports. flush() returns after the last byte has been sent, and txdone() also reports True after the last byte has been sent.

This required to add two very simple functions down the stack to uart.c and ra.sci.c.

- One for telling, whether the transmission is busy.
- One for reporting the size of the TX buffer. There is no txbuf option for setting the bufger size, so the actually used buffer size had to be retrieved.

Tested with a EK-RA6M2 board by me and with all boards from the port's boards directory by @TakeoTakahashi2020.

### CODE REVIEW

File: ports/renesas-ra/machine_uart.c [hunk 1]
```diff
@@ -409,6 +409,14 @@ STATIC mp_obj_t machine_uart_sendbreak(mp_obj_t self_in) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_uart_sendbreak_obj, machine_uart_sendbreak);
 
+/// \method uart.txdone()
+/// Return `True` if all characters have been sent.
```
Comment (dpgeorge) [inline ports/renesas-ra/machine_uart.c:413]: I know it's a bit consistent in this file, but please use only double slash `//` to start these comments.

File: ports/renesas-ra/machine_uart.c [hunk 2]
```diff
@@ -557,6 +567,20 @@ STATIC mp_uint_t machine_uart_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr
...
             ret |= MP_STREAM_POLL_WR;
         }
+    } else if (request == MP_STREAM_FLUSH) {
+        // The timeout is estimated using the buffer size and the baudrate.
+        // Take the worst case assumptions at 13 bit symbol size times 2.
+        // t.b.d. get actial TX buffer size
+        uint32_t timeout = mp_hal_ticks_ms() +
+            (uint32_t)(uart_tx_txbuf(self)) * 13000ll * 2 / self->baudrate;
+        do {
+            if (machine_uart_txdone((mp_obj_t)self) == mp_const_true) {
```
Comment (dpgeorge) [inline ports/renesas-ra/machine_uart.c:577]: Why not just call `!uart_tx_busy(self)` directly?  That will be a tiny bit faster, and not require a cast.

If you need to cast, you should use `MP_OBJ_FROM_PTR(self)` (that's compatible with nan-boxing mode).

**Review by dpgeorge — CHANGES_REQUESTED** [review]

File: ports/renesas-ra/machine_uart.c (hunk 2 above)
Comment (robert-hh) [reply ports/renesas-ra/machine_uart.c:577]: Thanks for the hint. Done. I'll re-test and update.

Comment (robert-hh) [reply ports/renesas-ra/machine_uart.c]: The changes are tested and ammended to the commit, which is updated.

Comment (dpgeorge) [reply ports/renesas-ra/machine_uart.c]: Excellent!

File: ports/renesas-ra/machine_uart.c [hunk 3]
```diff
@@ -557,6 +567,20 @@ STATIC mp_uint_t machine_uart_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr
         if ((flags & MP_STREAM_POLL_WR) && uart_tx_avail(self)) {
             ret |= MP_STREAM_POLL_WR;
         }
+    } else if (request == MP_STREAM_FLUSH) {
+        // The timeout is estimated using the buffer size and the baudrate.
+        // Take the worst case assumptions at 13 bit symbol size times 2.
+        // t.b.d. get actial TX buffer size
```
Comment (dpgeorge) [inline ports/renesas-ra/machine_uart.c:573]: Can this comment line be removed, since the actual TX buffer size is being retrieved via `uart_tx_txbuf`?

Comment (dpgeorge) [reply ports/renesas-ra/machine_uart.c:573]: (if the line stays, typo fix actial -> actual)

Comment (robert-hh) [reply ports/renesas-ra/machine_uart.c:573]: Sure. This can go. I'll remove the last line of the comment.

Comment (robert-hh) [reply ports/renesas-ra/machine_uart.c]: ... and updated.

**Comment by dpgeorge** [comment]
Thanks for updating.  Merged as-is.

**Comment by robert-hh** [comment]
Thanks for this mid-night merge.

**Pull request merged**